src/hotspot/share/utilities/rbTree.hpp

```diff
@@ -0,0 +1,313 @@
+/*
+ * Copyright (c) 2025, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_UTILITIES_RBTREE_HPP
+#define SHARE_UTILITIES_RBTREE_HPP
+
+#include "nmt/memTag.hpp"
+#include "runtime/os.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include <type_traits>
+
+// COMPARATOR must have a static function `cmp(a,b)` which returns:
+//     - an int < 0 when a < b
+//     - an int == 0 when a == b
+//     - an int > 0 when a > b
+// ALLOCATOR must check for oom and exit, as RBTree currently does not handle the
+// allocation failing.
+// Key needs to be of a type that is trivially destructible.
+// The tree will call a value's destructor when its node is removed.
+// Nodes are address stable and will not change during its lifetime.
+
+template <typename K, typename V, typename COMPARATOR, typename ALLOCATOR>
+class RBTree {
+  friend class RBTreeTest;
+
+private:
+  ALLOCATOR _allocator;
+  size_t _num_nodes;
+
+public:
+  class RBNode {
+    friend RBTree;
+    friend class RBTreeTest;
+
+  private:
+    RBNode* _parent;
+    RBNode* _left;
+    RBNode* _right;
+
+    const K _key;
+    V _value;
+
+    enum Color : uint8_t { BLACK, RED };
+    Color _color;
+
+  public:
+    const K& key() const { return _key; }
+    V& val() { return _value; }
+
+  private:
+    bool is_black() const { return _color == BLACK; }
+    bool is_red() const { return _color == RED; }
+
+    void set_black() { _color = BLACK; }
+    void set_red() { _color = RED; }
+
+    RBNode(const K& k, const V& v)
+        : _parent(nullptr), _left(nullptr), _right(nullptr),
+          _key(k), _value(v), _color(RED) {}
+
+    bool is_right_child() const {
+      return _parent != nullptr && _parent->_right == this;
+    }
+
+    bool is_left_child() const {
+      return _parent != nullptr && _parent->_left == this;
+    }
+
+    void replace_child(RBNode* old_child, RBNode* new_child);
+
+    // Move node down to the left, and right child up
```

Comment:
```suggestion
    // This node down, right child up
    // Returns right child (now parent)
```
(different papers use left/right in different ways, I always get confused. A clear concise comment helps)